File: solutions/elixir/raindrops/1/lib/raindrops.ex
defmodule Raindrops do
  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    result =
      [3, 5, 7]
      |> Enum.reduce("", fn divisor, acc ->
        if rem(number, divisor) == 0 do
          case divisor do
            3 -> acc <> "Pling"
            5 -> acc <> "Plang"
            7 -> acc <> "Plong"
          end
        else
          acc
        end
      end)

    case result do
      "" -> Integer.to_string(number)
      _ -> result
    end
  end
end

File: solutions/cpp/raindrops/1/raindrops.cpp
#include "raindrops.h"
#include <string>

namespace raindrops {

std::string convert(int n) {
  std::string result = "";

  if (n % 3 == 0)
    result.append("Pling");
  if (n % 5 == 0)
    result.append("Plang");
  if (n % 7 == 0)
    result.append("Plong");

  if (result.length() == 0)
    return std::to_string(n);

  return result;
}

} // namespace raindrops
